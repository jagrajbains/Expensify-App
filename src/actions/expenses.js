import uuid from "uuid";
import database from "../firebase/firebase";

//     1. ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});
/**
 * By default redux doesn't allow us to export a function
 * Since we have integrated redux-thunk and we are using the thunkMiddleware
 * we will be able to dispatch a callback function
 */
export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/Expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};
//     2. REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});
//     3. EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

//SET_EXPENSES
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

//START_SET_EXPENSES
export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/Expenses`)
      .once("value")
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};

//startRemoveExpenses
export const startRemoveExpenses = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/Expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(
          removeExpense({
            id
          })
        );
      });
  };
};

//startEditExpense
export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/Expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};
