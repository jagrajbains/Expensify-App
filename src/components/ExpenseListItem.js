import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
import "../locales/inr";
const ExpenseListItem = ({ id, description, amount, createdAt }) => {
  numeral.locale("inr");
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        {numeral(amount).format("$0,0.00")}-
        {moment(createdAt).format("ddd, MMMM Do YYYY")}
      </p>
    </div>
  );
};

export default ExpenseListItem;
