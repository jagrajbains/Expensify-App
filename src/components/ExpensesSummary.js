import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import "../locales/inr";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expense-total";

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  numeral.locale("inr");
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  const formattedExpensesTotal = numeral(expensesTotal).format("$0,0.00");
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
      </h1>
    </div>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
