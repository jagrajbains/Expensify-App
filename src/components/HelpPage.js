import React from 'react';

const HelpPage = () => (
    <div>
        <p>This is from the help Page</p>
    </div>
);

export default HelpPage;