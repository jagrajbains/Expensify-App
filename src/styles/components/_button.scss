.button {
  background-color: $blue;
  border-radius: 4px;
  border: none;
  color: #fff;
  display: inline-block;
  font-weight: 300;
  line-height: 1;
  padding: $s-size $m-size;
  text-decoration: none;
}

.button--link {
  background: none;
}

.button--danger {
  @extend .button;
  background-color: #d9534f;
}
