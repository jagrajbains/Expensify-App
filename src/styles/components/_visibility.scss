.show-for-mobile {
  @media (min-width: $desktop-breakpoint) {
    display: none;
  }
}

.show-for-largerScreens {
  @media (max-width: $desktop-breakpoint - 0.01rem) {
    display: none;
  }
}
