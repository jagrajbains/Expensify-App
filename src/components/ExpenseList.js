import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectedExpenses from "../selectors/expenses";

export const ExpenseList = props => {
  return (
    <div className="content-container">
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-largerScreens">Expense</div>
        <div className="show-for-largerScreens">Amount</div>
      </div>
      <div className="list-body">
        {props.expenses.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Expenses Added.</span>
          </div>
        ) : (
          props.expenses.map(expense => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )}
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: selectedExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
