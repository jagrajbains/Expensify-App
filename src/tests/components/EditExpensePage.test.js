import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let wrapper, history, startEditExpense, startRemoveExpenses;
beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpenses = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      expense={expenses[2]}
      startEditExpense={startEditExpense}
      startRemoveExpenses={startRemoveExpenses}
      history={history}
    />
  );
});

test("Should render EditExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle startEditExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startEditExpense).toHaveBeenLastCalledWith(
    expenses[2].id,
    expenses[2]
  );
});

test("should handle startRemoveExpenses", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startRemoveExpenses).toHaveBeenLastCalledWith({
    id: expenses[2].id
  });
});
