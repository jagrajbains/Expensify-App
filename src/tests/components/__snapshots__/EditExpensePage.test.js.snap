// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 2500,
          "createdAt": 345600000,
          "description": "Credit Card Bill",
          "id": "3",
          "note": "This month's credit card bill",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button--danger"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
